### PR TITLE
[libc] [unistd] implement pipe2 syscall wrapper

### LINKED ISSUE #85289: [libc] implement pipe2
The pipe2 is defined on linux here: https://man7.org/linux/man-pages/man2/pipe.2.html

This is a simple syscall wrapper, so implementation should be straightforward.

### Build System

In [entrypoints.txt](https://github.com/llvm/llvm-project/blob/main/libc/config/linux/x86_64/entrypoints.txt)
- [ ] Add the new function to the list of entrypoints for unistd.h (you need to do this before it will build)

In [unistd/CMakeLists.txt](https://github.com/llvm/llvm-project/blob/main/libc/src/unistd/CMakeLists.txt)
- [ ] Add new entrypoint objects to the list
- [ ] Also add it to the list for unistd/linux/CMakeLists.txt

In [linux.td](https://github.com/llvm/llvm-project/blob/main/libc/spec/linux.td)
- [ ] Add the functions to the list of functions for unistd.h (this can be confusing, ask for help if you're stuck).

### Code

Create `unistd/pipe2.h`
- [ ] look at [dup.h](https://github.com/llvm/llvm-project/blob/main/libc/src/unistd/dup.h) for an example of how this should be laid out (copying the whole file and modifying it is fine).

Create `unistd/linux/pipe2.cpp`
- [ ] Again, look at look at [linux/dup.cpp](https://github.com/llvm/llvm-project/blob/main/libc/src/unistd/linux/dup.cpp) for an example.

### Testing

Create `test/src/unistd/pipe2_test.cpp`
- [ ] This needs to test the function succeeding and failing
- [ ] This is only testing the code you wrote, so don't worry too much about testing parts of the interface that the kernel handles.

Add your tests to [test/src/unistd/CMakeLists.txt](https://github.com/llvm/llvm-project/blob/main/libc/test/src/unistd/CMakeLists.txt)
- [ ] Hopefully self-explanatory

### PR DESCRIPTION
**Pull request by duncpro**
Closes #85289

### CODE REVIEW

File: libc/test/src/unistd/pipe2_test.cpp
Comment (nickdesaulniers) [inline libc/test/src/unistd/pipe2_test.cpp:30]: The tests aren't as thorough as https://github.com/llvm/llvm-project/pull/85514/files, but I think that's ok for syscall wrappers.

**Comment by duncpro** [comment]
Just noticed `spec/linux.td` should declare `pipe2` not `pipe`. I need to fix before we can merge.

File: libc/spec/linux.td
Comment (nickdesaulniers) [inline libc/spec/linux.td:298]: ah, good catch, yeah this should be `pipe2`. Sorry I missed it!

File: libc/test/src/unistd/pipe2_test.cpp
Comment (michaelrj-google) [reply libc/test/src/unistd/pipe2_test.cpp:30]: syscall wrapper tests just need to exercise the libc side, so for most syscalls we just need a success and an error. It's also good to check that the values it returns are valid, since it's easy to forget to unpoison memory for sanitizers.

File: libc/src/unistd/linux/pipe2.cpp
Comment (michaelrj-google) [inline libc/src/unistd/linux/pipe2.cpp:22]: You need to unpoison the memory here for the sanitizers to work properly. They can't tell that the kernel initialized the memory without being explicitly told.
```suggestion
  MSAN_UNPOISON(pipefd, sizeof(int) * 2);
```

**Review by michaelrj-google — COMMENTED** [review]
overall LGTM, with some minor tweaks. Thanks for the patch!

File: libc/test/src/unistd/pipe2_test.cpp
Comment (duncpro) [reply libc/test/src/unistd/pipe2_test.cpp:30]: The tests at https://github.com/llvm/llvm-project/pull/85514/files seem to be testing the kernel itself. For instance they verify that it's possible to read/write to the file descriptors, and that the descriptors are unique, etc. AFAIK this is outside the scope of libc. We assume the kernel is correct.

As far as the tests in the new PR go,
- We've verified that when return code is `-1`, `libc_errno` is populated.
- We've verified that `pipefd` holds valid file descriptors when `pipe2` returns 0. (If they were invalid [`close`](https://www.man7.org/linux/man-pages/man2/close.2.html) would not be successful.)

Is there anything I'm missing here?

File: libc/spec/linux.td
Comment (duncpro) [reply libc/spec/linux.td]: No worries! This is fixed in [b1e2ea6](https://github.com/llvm/llvm-project/commit/b1e2ea64182ddab6775ab6cfe5168c98c92c1a57).

File: libc/test/src/unistd/pipe2_test.cpp
Comment (nickdesaulniers) [reply libc/test/src/unistd/pipe2_test.cpp:30]: > Can it be removed here?

Sure. The linux kernel contains kselftests to exercise numerous things, including the syscall interfaces.  Ideally, the functional test suite we'd use is to build the Linux kernel's kselftests against llvm-libc, then run the result.

**Review by nickdesaulniers — APPROVED** [review]

**Comment by nickdesaulniers** [comment]
Do you need us to merge this for you? If not, please wait until the presubmits complete and are green before merging.

**Comment by duncpro** [comment]
I don't think I have write access

**Comment by nickdesaulniers** [comment]
ok, will merge once green. ping this thread if it's not merged by monday afternoon

File: libc/spec/linux.td
Comment (michaelrj-google) [inline libc/spec/linux.td:300]: This needs to be `[ArgSpec<IntPtr>, ArgSpec<IntType>]`. Looks like it's also wrong for `pipe`, but that can be fixed later.
```suggestion
          [ArgSpec<IntPtr>, ArgSpec<IntType>] //TODO: make this int[2]
```

**Review by michaelrj-google — COMMENTED** [review]
Noticed one small issue in the oldhdrgen files.

File: libc/spec/linux.td
Comment (duncpro) [reply libc/spec/linux.td]: Sorry about that. 

 Is there an automated test I should be running which catches discrepancies like this? Or do we just have to be careful here.

Comment (duncpro) [reply libc/spec/linux.td]: Also, AFAIK the `pipe` definition in `posix.td` actually doesn't have this problem.

https://github.com/llvm/llvm-project/blob/f54cdc5d6ee5532da117f2489c105148c94dcb39/libc/spec/posix.td#L646-L650

The [posix standard for `pipe`](https://pubs.opengroup.org/onlinepubs/9799919799/basedefs/unistd.h.html) defines the signature as...

```
int pipe(int[2]);
```
This matches the `FunctionSpec` we have in `posix.td`.

Comment (nickdesaulniers) [reply libc/spec/linux.td]: @petrhosek is working on removing the old headergen implementation; we expect it to all go away imminently. So mistakes in the existing TD files are not pressing.

**Comment by duncpro** [comment]
@nickdesaulniers Can we merge this?

File: libc/src/unistd/linux/pipe2.cpp
Comment (michaelrj-google) [inline libc/src/unistd/linux/pipe2.cpp:26]: nit: The unpoison should go after the `if(ret < 0>{}` so that the memory is only considered initialized if the kernel has actually written to it.
```suggestion
  if (ret < 0) {
    libc_errno = -ret;
    return -1;
  }
  MSAN_UNPOISON(pipefd, sizeof(int) * 2);
```

**Review by michaelrj-google — APPROVED** [review]
LGTM with one minor nit

File: libc/src/unistd/linux/pipe2.cpp
Comment (duncpro) [reply libc/src/unistd/linux/pipe2.cpp]: Fixed in a71f550

**Review by michaelrj-google — APPROVED** [review]
LGTM, thanks for the patch! I'll let Nick merge this.

**Pull request merged**
